### DISCUSSION TITLE
Wrapper / Template for non-Typescript models

### DISCUSSION THREAD

**Discussion by pazbardanl** [Green-Software-Foundation/hack#36] [Develop an Idea]
## Prize category

- [x] Best Model
- [ ] Best Content
- [ ] Beyond Carbon
- [ ] Best contribution to the framework

## User Story
- As a **IF model developer**, I want **to develop non-TypeScript models starting with a wrapper/template** so that **I can focus on my custom business logic** rather than dealing with the generic / boilerplate stuff.

## Rationale
Developers in the community may choose to build models using other languages other than TypeScript. For a given language, the (possibly) many models written have different business logic internally, but they share a lot of common code related to IF interactions:
- SDTIN / SDTOUT handling
- Parsing certain fields in the incoming impl/yml and passing them to the business logic
- Appending the results onto the incoming inputs to create an output block
- Input validation

## Impact
If this idea is successfully implemented the creation of non-Typescript models will be significantly faster. Developers will need not worry about handling of input, composing the output and communicating over STDIN/STDOUT. They could focus on their custom business logic and deliver models faster.

## Implementation guidelines
* This model MUST conform to the `ModelPlugin` interface
* This model can be one the following: (1) a **template**: piece of code with a blank space one needs to fill with their custom model business logic (2) a **Wrapper**: a ready-to-use library to which one passes a reference to their code. IF shell model calls the wrapper that in turn calls the user's code.
* This model can be in any programming language/tech, although the more commonly used it is the better. https://octoverse.github.com/2022/top-programming-languages
* Hackers are encouraged to: (1) Build more than 1 model (for more than 1 language). Once one is done, the other may take significantly less time to build (2) add in more "frameworky" stuff for consistency such as a unified error message format

## Difficulty
- If we provide an example for one the languages: 3/5
- if no example is provided: 4/5

## Scope Of Work

- [x] A new model plugin + **documentation**
- [ ] A new feature in an existing model plugin
- [ ] A manifest file will have to be written
- [ ] Implement test cases
- [ ] A new feature to the framework itself (the cli tool)
- [ ] It will require changes to the documentation

## Examples and resources
Shell model: https://github.com/Green-Software-Foundation/if-models/blob/main/src/lib/shell/README.md

## Manifest File
N/A

**Comment by jmcook1186** [comment]
Hi @pazbardanl - intriguing idea. Can you explain in any more detail why this would be an improvement over our existing `ShellModel` plugin?

**Reply by pazbardanl** [reply]
Hi @jmcook1186 
The shell model is fine as it is, my focus with this idea is the piece of code it invokes: for a given language, if you look at all these scripts that are invoked by the shell model you'll find they share a lot of boilerplate stuff (stdin/out handling, parsing yaml objects, input validation etc) and only small portion of it is actual model logic, which is of course different from one script to another.
So I'm saying: it'll be more efficient to decouple boilerplate and core business logic and provide developers the former (which is common for all) as a wrapper or template (or some other type of implementation) so that they can focus on the latter.
Hope I've managed to make it clearer 🙂

**Reply by jmcook1186** [reply]
makes sense - thanks!

**Reply by jawache** [reply]
Thanks, @pazbardanl; we do need to make it easier for people to write plugins in different languages. Is this something per language?

So a wrapper/boiler plate for say, Java, which makes building Java plugins easier. And another wrapper/plugin for say C# that makes writing C# easier. Something like that?

**Reply by pazbardanl** [reply]
@jawache yeah that's the plan. At first glance it may seem overwhelming, a wrapper for each language.. BUT we don't really need one for EVERY language, just for the most popular ones. and this why i think the hackathon is an opportunity: we'll have lots of devs out there, and whatever languages they choose to build the wrappers are the ones in demand by actual ppl in our community.
Plus - the potential for collaboration here is pretty big: imagine 5 devs building 5 wrappers in 5 languages and submitting them together.

**Reply by jawache** [reply]
Hmm if we are talking wrappers, then perhaps take it a step further and talk about boilerplate code that creates native language bindings.

With nodejs you can bind to c/CPP (https://nodejs.org/api/n-api.html#node-api) so call c directly from nodejs.

Most languages have some way to expose a c interface since its so low level so we end up being able to natively call any c-linkable language directly from TS, as if it's a function call!

Or was that basically what you were suggesting all along @pazbardanl ? 😂

**Reply by pazbardanl** [reply]
Well @jawache  TBH that's not my original suggestion but it's interesting nonetheless :)

Just to make sure I'm following your thread: you're saying most non-TS 'plugins' (= code to be actually invoked by the shell plugin) will be written in languages that expose a C interface (such as [cython](https://cython.org/) ?), so let's make that the standard instead of the STDIN/OUT communication?

If so then i think it's good to support C interfaces (and a good hackathon idea as well), but need to understand if this method of writing non-TS plugins can act as replacement to writing code that handles STDIN/OUT. I personally have no experience with C interfaces, but I will do some research as to how popular this method is.

The main thing I wanted to achieve with this idea is to have something that **encourages** non-TS developers (like me until recently 😸) to write plugins. One that's great about the STDIN/OUT "protocol" is that it **completely** decouples the shell plugin from the code it is invoking, it doesn't even know it's there. I fear having our TS/nodejs shell model call C directly means it becomes "aware" of how it is implemented which kinda breaks this perfect decoupling.

**Reply by jawache** [reply]
Oh I see. So the boilerplate code you were talking about is boilerplate code for different languages handling standard in and standard out?

If so, totally agree. That'll make everybody's lives a lot easier.

I do think we need to explore a different approach. Currently, the shell model plug-in is instantiated for every single component in a graph separately, which means you call configure and then pass to standard in, and then end the process for every component. 

We should explore a way where you start the process when you start running the manifest file and then each component gets passed in as a separate standard in and then there's a protocol where the shell passes something out so that the impact engine knows that it's finished processing without ending the process.

Hope I'm making sense?

**Reply by pazbardanl** [reply]
"_Oh I see. So the boilerplate code you were talking about is boilerplate code for different languages handling standard in and standard out?_" -- **Yes**, but it's not only STDIN/OUT, it's also yml parsing, input validation and output yml handling (and maybe other aspects as well) that are common for all plugins and can (and should, in my view) be decoupled from the business logic.

Regarding the different approach you described:
If I understand correctly the issues (and solutions) we are talking about are 2 separate ones:

(a) Improving development efficiency of "external scripts" (=code to be invoked by the shell plugin) by providing boilerplate code or a wrapper for a handful of selected languages (the more languages hackers tend to the most impressive their submission will be, i can only assume)
(b) Improving the performance of the shell plugin by executing a single instance of it to handle all components, rather than executing it once per component.

if i understand correctly these are 2 separate issues to which the solutions could be developed orthogonally, hence both may be their own hackathon ideas. right?

**Reply by jawache** [reply]
Correct, it's just that (a) might change depending on (b) but saying that.... we are figuring (b) out in the IF refactor right now so in 2 weeks we'll know for sure.

**Reply by pazbardanl** [reply]
hey @jawache, @narekhovhannisyan , @jmcook1186 
With the design of the new framework now known, i think we may be in a good place to make a decision regarding this.
in my view (a) and (b) could be 2 separate ideas, most likely orthogonal.
let me know what you think.

**Reply by NAMRATA-WOKE** [reply]
@jawache @jmcook1186 pinging you again to close the loop.
